app/src/main/java/org/mozilla/fenix/library/history/HistoryView.kt

```diff
@@ -31,8 +33,25 @@ class HistoryView(
     var mode: HistoryFragmentState.Mode = HistoryFragmentState.Mode.Normal
         private set
 
-    val historyAdapter = HistoryAdapter(interactor)
+    val historyAdapter = HistoryAdapter(interactor).apply {
+        addLoadStateListener {
+            // first call will always have itemCount == 0, but we want to keep adapterItemCount
+            // as null until we can distinguish an empty list from populated, so updateEmptyState()
+            // could work correctly
+            if (itemCount != 0) {
+                adapterItemCount = itemCount
+            }
+            if (it.source.refresh is LoadState.NotLoading &&
+                it.append.endOfPaginationReached &&
+                itemCount < 1
+            ) {
+                adapterItemCount = 0
+                onZeroItemsLoaded.invoke()
```

Comment:
I can kind of see a path where we can remove the `userHasHistory` LiveData (we have been trying to remove LiveData afaik) and instead make use of the `HistoryFragmentStore`. Consequently,  `HistoryFragmentState` also has an unused `items` property, but you can easily dispatch an action to the `HistoryFragmentStore` to tell it whether or not we have any items and update the UI as appropriate. We can file this as a followup with the intent of removing `userHasHistory` LiveData and perhaps cleaning up `HistoryFragmentState`. I don't think we necessarily need to cache the list of history items, but maybe we can just have an equivalent to this `hasHistory` state there so that we can respond to any state changes without doing `model.userHasHistory.observe`.

Reply:
opened an issue
https://github.com/mozilla-mobile/fenix/issues/24434